kornia/enhance/jpeg.py

```diff
@@ -0,0 +1,573 @@
+    return output
+
+
+def _dct_8x8(input: Tensor) -> Tensor:
+    """Performs an 8 x 8 discrete cosine transform.
+
+    Args:
+        input (Tensor): Patched input tensor of the shape :math:`(B, N, 8, 8)`.
+
+    Returns:
+        output (Tensor): DCT output tensor of the shape :math:`(B, N, 8, 8)`.
+    """
+    # Get dtype and device
+    dtype: Dtype = input.dtype
+    device: Device = input.device
+    # Make DCT tensor and scaling
+    index: Tensor = torch.arange(8, dtype=dtype, device=device)
+    x, y, u, v = torch.meshgrid(index, index, index, index)
+    dct_tensor: Tensor = ((2.0 * x + 1.0) * u * torch.pi / 16.0).cos() * ((2.0 * y + 1.0) * v * torch.pi / 16.0).cos()
+    alpha: Tensor = torch.ones(8, dtype=dtype, device=device)
+    alpha[0] = 1.0 / (2**0.5)
+    dct_scale: Tensor = torch.einsum("i, j -> ij", alpha, alpha) * 0.25
+    # Apply DCT
+    output: Tensor = dct_scale[None, None] * torch.tensordot(input - 128.0, dct_tensor)
+    return output
+
+
+def _idct_8x8(input: Tensor) -> Tensor:
+    """Performs an 8 x 8 discrete cosine transform.
+
+    Args:
+        input (Tensor): Patched input tensor of the shape :math:`(B, N, 8, 8)`.
+
+    Returns:
+        output (Tensor): DCT output tensor of the shape :math:`(B, N, 8, 8)`.
+    """
+    # Get dtype and device
+    dtype: Dtype = input.dtype
+    device: Device = input.device
+    # Make and apply scaling
+    alpha: Tensor = torch.ones(8, dtype=dtype, device=device)
+    alpha[0] = 1.0 / (2**0.5)
+    dct_scale: Tensor = torch.outer(alpha, alpha)
+    input = input * dct_scale[None, None]
+    # Make DCT tensor and scaling
+    index: Tensor = torch.arange(8, dtype=dtype, device=device)
+    x, y, u, v = torch.meshgrid(index, index, index, index)
+    idct_tensor: Tensor = ((2.0 * u + 1.0) * x * torch.pi / 16.0).cos() * ((2.0 * v + 1.0) * y * torch.pi / 16.0).cos()
+    # Apply DCT
+    output: Tensor = 0.25 * torch.tensordot(input, idct_tensor, dims=2) + 128.0
+    return output
+
+
+def _differentiable_polynomial_rounding(input: Tensor) -> Tensor:
+    """This function implements differentiable rounding.
+
+    Args:
+        input (Tensor): Input tensor of any shape to be rounded.
+
+    Returns:
+        output (Tensor): Pseudo rounded tensor of the same shape as input tensor.
+    """
+    # Perform differentiable rounding
+    output: Tensor = torch.round(input) + (input - torch.round(input)) ** 3
+    return output
+
+
+def _differentiable_polynomial_floor(input: Tensor) -> Tensor:
+    """This function implements differentiable floor.
+
+    Args:
+        input (Tensor): Input tensor of any shape to be floored.
+
+    Returns:
+        output (Tensor): Pseudo rounded tensor of the same shape as input tensor.
+    """
+    # Perform differentiable rounding
+    output: Tensor = torch.floor(input) + (input - 0.5 - torch.floor(input)) ** 3
+    return output
+
+
+def _differentiable_clipping(
```

Comment:
for a further pr, i think having `_differentiable_clipping`, ` _differentiable_polynomial_floor` and ` _differentiable_polynomial_rounding` into our utilities can be usefully :)

Reply:
open a ticket for that